### PR TITLE
Early constructor resolution & import-loops immune compiler

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Equips the compiler with the possibility to work on multiple modules at once (possibly with loops between them).
Implements a few passes using the newly acquired capabilities to perform early resolution of constructors in method definitions and patterns, thus removing this logic from Code Gen's responsibilities.

### Important Notes
This is part 1 of the new module system.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Review by iamrecursion — CHANGES_REQUESTED** [review]
An amazing load of work, but there are some changes I'd like to see.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java [hunk 1]
```diff
@@ -33,23 +36,42 @@
...
 public class Module implements TruffleObject {
+
+  /** Defines a stage of compilation of the module. */
+  public enum CompilationStage {
+    INITIAL(0),
+    AFTER_PARSING(1),
+    AFTER_IMPORT_RESOLUTION(2),
+    AFTER_STATIC_PASSES(3),
+    AFTER_RUNTIME_STUBS(4),
+    AFTER_CODEGEN(5);
```
[the remark this answers is not in the document]
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:47]: > I would only set the number for INITIAL and let the rest auto-increment

This is how C++ works, not Java. There are 2 options:
1. Have a constructor field and use it explicitly (implemented here).
2. Depend on definition order implicitly.

Solution 2. Seems bad to me, as there is nothing signifying "this enum is in fact ordered".

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala [hunk 2]
```diff
@@ -2,67 +2,132 @@ package org.enso.compiler
...
   *
   * @param context the language context
   */
-class Compiler(private val context: Context) {
+class Compiler(val context: Context) {
   private val freshNameSupply: FreshNameSupply = new FreshNameSupply
   private val passes: Passes                   = new Passes
   private val passManager: PassManager         = passes.passManager
+  private val importResolver: ImportResolver   = new ImportResolver(this)
+  private val stubsGenerator: TruffleStubsGenerator =
```
[the remark this answers is not in the document]
Comment (kustosz) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala]: It does not generate nodes at all (the whole point is: generate empty shells for structures, without populating the nodes yet). How about `RuntimeStubsGenerator`?

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        val ir = current.getIr
+        val currentLocal = ir.unsafeGetMetadata(
+          BindingResolution,
+          "Non-parsed module used in ImportResolver"
+        )
+        val importedModuleNames = ir.imports.collect {
+          case imp: IR.Module.Scope.Import.Module => imp.name
+        }
+        val importedModules = importedModuleNames.flatMap(compiler.getModule)
+        // TODO[MK] Remove when No Implicit Prelude
```
[the remark this answers is not in the document]
Comment (kustosz) [reply engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:55]: Yes

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (iamrecursion) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:47]: Gross, but fine. Leave as is!

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamrecursion) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala]: That works.

File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/PatternResolution.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  /** Executes the pass on the provided `ir`, and returns a possibly transformed
+    * or annotated version of `ir` in an inline context.
+    *
+    * @param ir            the Enso IR to process
+    * @param inlineContext a context object that contains the information needed
+    *                      for inline evaluation
+    * @return `ir`, possibly having made transformations or annotations to that
+    *         IR.
+    */
+  override def runExpression(
```
[the remark this answers is not in the document]
Comment (kustosz) [reply engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/PatternResolution.scala]: But... it is

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**